File: android/app/src/main/kotlin/com/example/hotel_reception/procesor/model/Hotel.kt
package com.example.hotel_reception.procesor.model

import android.util.Log
import com.example.hotel_reception.procesor.WorldsClock
import com.example.hotel_reception.procesor.dto.WorldSpeed
import com.example.hotel_reception.procesor.event.EventGenerator
import com.example.hotel_reception.procesor.staff.AccountantService
import com.example.hotel_reception.procesor.staff.HotelReception
import com.example.hotel_reception.procesor.staff.HotelService
import com.example.hotel_reception.procesor.staff.RepairingSquad
import java.util.*

const val logTag = "HotelService"

class Hotel(
    private val service: Service = Service.normal,
    private val _rooms: List<HotelRoom>,
    private val worldsClock: WorldsClock,
    private var _balance: Int = 5000,
) : HotelService, AccountantService, HotelReception, RepairingSquad {
    val rooms: List<HotelRoom> get() = _rooms
    val balance: Int get() = _balance
    val monthRevenue: Int get() = _monthRevenue
    val dayRevenue: Int get() = _dayRevenue

    val today: Date get() = worldsClock.getDate()
    val currentSpeed: WorldSpeed = worldsClock.getCurrentSpeed()

    private var _monthRevenue = 0
    private var _dayRevenue = 0

    companion object {
        fun buildHotel(
            service: Service,
            roomByType: List<RoomByType>,
            worldsClock: WorldsClock
        ): Hotel {

            val rooms = mutableListOf<BasicRoom>()
            roomByType.forEach { type ->
                rooms.addAll(List(type.count) {
                    return@List BasicRoom(
                        roomType = type.roomType,
                        number = rooms.size + it
                    )
                })
            }
            return Hotel(
                service,
                _rooms = rooms.map { return@map HotelRoom(it) },
                worldsClock
            )
        }
    }

    override fun checkDesk() {
        decreaseRoomLifeTime()
        cleanRooms()
        calculateOutCome()
        checkoutCustomers()
        renewRoomsAfterRepairing()
        checkRoomsNeedRepairing()
    }


    override fun calculateIncome(inCome: Int) {
        _dayRevenue = inCome
        _monthRevenue += _dayRevenue
        _balance += _dayRevenue
        Log.i(
            logTag,
            "income $inCome"
        )
    }


    override fun calculateOutCome() {
        var outCome = 0
        outCome += service.salary
        val roomsInUseCount = _rooms.filter(HotelRoom::occupiedByCustomer).size
        outCome += roomsInUseCount * 20 + _rooms.size - roomsInUseCount
        _balance -= outCome
        Log.i(logTag, "outcome $outCome")
    }

    override fun payTaxes() {
        val tax = _monthRevenue * .1
        _balance -= tax.toInt()
        _monthRevenue = 0
        Log.i(logTag, "taxes paid, value: $tax")
    }

    override fun cleanRooms() {
        val roomsInService = _rooms.filter(HotelRoom::roomWaitingClearing)
        val indexes = roomsInService.take(service.roomPerDay).map { it.room.number }
        Log.i(
            logTag,
            "clearRoom, rooms in progress ${indexes.size}, rooms currently in service ${roomsInService.size}"
        )

        indexes.forEach {
            _rooms[it].roomBusiness = EmptyRoom()
        }
    }

    override fun checkoutCustomers() {
        var checkedOutCustomers = 0
        _rooms.forEach {
            if (it.occupiedByCustomer) {
                if ((it.roomBusiness as RoomInUse).endDate == worldsClock.getDate()) {
                    checkedOutCustomers++
                    it.roomBusiness = DirtyRoom()
                }
            }
        }
        Log.i(
            logTag,
            "checkDesk ${worldsClock.getDate()}, customers checked out: $checkedOutCustomers"
        )
    }

    override fun checkInPersons(persons: List<Person>) {
        val freeRooms = _rooms.filter(HotelRoom::readyForCustomer)
        Log.i(
            logTag,
            "newPersonsArrived persons count ${persons.size} free rooms ${freeRooms.size}"
        )
        val roomsByType = freeRooms.groupBy { it.room.roomType }.toList().sortedBy {
            return@sortedBy when (it.first) {
                RoomType.president -> 1
                RoomType.suite -> 2
                else -> 3
            }
        }
        var inCome = 0
        roomsByType.forEach { roomByType ->
            val personsByPreferences =
                persons.filter { it.roomPreferences.contains(roomByType.first) }
            val freeRoomsByType = roomByType.second
            if (freeRoomsByType.isNotEmpty() && personsByPreferences.isNotEmpty()) {
                val personToCheckIn = personsByPreferences.take(freeRoomsByType.size)
                personToCheckIn.forEachIndexed { index, person ->
                    inCome += (freeRoomsByType[index].room.roomType.cost * worldsClock.getSeason().priceMultiply).toInt() * personToCheckIn[index].daysAmount
                    _rooms[freeRoomsByType[index].room.number]
                    _rooms[freeRoomsByType[index].room.number].also { room ->
                        room.roomBusiness = CustomerRoom(
                            personToCheckIn[index].startDate,
                            personToCheckIn[index].endDate
                        )
                        EventGenerator.breakEventGenerated(person)?.let {
                            if (it.completelyBreak) {
                                Log.i(
                                    logTag,
                                    "---------------------person completely break------------------"
                                )
                                room.lifetime = 0
                            } else {
                                Log.i(
                                    logTag,
                                    "---------------------person damaged room by ${it.costToRepair} ------------------"
                                )
                                Log.i(
                                    logTag,
                                    "---------------------person damaged room by ${it.daysDecrement} days------------------"
                                )

                                _balance -= it.costToRepair
                                room.lifetime -= it.daysDecrement
                            }
                        }

                    }
                }
            }
        }
        calculateIncome(inCome)
    }

    override fun checkRoomsNeedRepairing() {
        val needRepair =
            _rooms.filter { it.lifetime < 1 && !it.occupiedByCustomer && !it.repairingRoom }
        Log.i(logTag, "checkRoomsNeedRepairing, need repair count ${needRepair.size}")
        needRepair.forEach(::repairRoom)
    }

    override fun renewRoomsAfterRepairing() {
        Log.i(
            logTag,
            "renewRoomsAfterRepairing, rooms in repairing ${_rooms.filter(HotelRoom::repairingRoom)}"
        )

        val renewedRooms =
            _rooms.filter { it.repairingRoom && (it.roomBusiness as RoomInUse).endDate == worldsClock.getDate() }
        Log.i(logTag, "renewRoomsAfterRepairing, renewed rooms ${renewedRooms.size}")


        renewedRooms.forEach {
            it.roomBusiness = EmptyRoom()
            it.lifetime = 90
        }
    }

    override fun decreaseRoomLifeTime() {
        Log.i(logTag, _rooms.map { it.lifetime }.joinToString())
        _rooms.filter(HotelRoom::occupiedByCustomer).forEach { it.lifetime-- }

        Log.i(logTag, "after decrease" + _rooms.map { it.lifetime }.joinToString())

    }

    override fun repairRoom(room: HotelRoom) {
        Log.i(logTag, "room became repairing, room number ${room.room.number}")
        _balance -= room.room.roomType.repairCost
        room.roomBusiness = RepairingRoom(
            worldsClock.getDate(),
            Calendar.getInstance().apply {
                time = worldsClock.getDate()
                add(Calendar.DATE, 5)
            }.time
        )
    }


}

enum class Service(val roomPerDay: Int, val salary: Int) {
    normal(5, 50), medium(10, 100), expensive(15, 150),
}

File: ios/Runner/Processor/model/hotel/Hotel.swift
//
//  Hotel.swift
//  Runner
//
//  Created by Serhii Rakovskyi on 18.08.2022.
//

import Foundation

class Hotel : HotelService, AccountantService, HotelReception, RepairingSquad {
    
    
    private let service:Service
    private var rooms: Array<HotelRoom>
    private var balance: Int = 5000
    private let worldsClock: WorldsClock
    private var monthRevenue: Int = 0
    private var dayRevenue: Int = 0
    
    static func build(service: Service,
                      roomByType: Array<RoomByType>,
                      worldsClock: WorldsClock)->Hotel {
        var rooms: Array<BasicRoom> = []
        
        roomByType.forEach { room in
            for i in 0...room.count {
                rooms.append(BasicRoom(roomType: room.roomType, number: rooms.count+i))
            }
        }
        return Hotel(service: service, rooms: rooms.map {
            HotelRoom.init(room: $0)}, worldsClock: worldsClock)
    }
    
    init(service:Service, rooms: Array<HotelRoom>, worldsClock: WorldsClock) {
        self.rooms = rooms
        self.service = service
        self.worldsClock = worldsClock
    }
    
    func getToday() -> Date{
        return worldsClock.getDate();
    }
    
    func getWorldsClock() -> WorldsClock{
        return worldsClock
    }
    
    func getMonthRevenue() -> Int{
        return monthRevenue
    }
    
    func getDayRevenue() -> Int {
        return dayRevenue
    }
    func getBalance() -> Int {
        return balance
    }
    
    func getRooms() -> Array<HotelRoom> {
        return rooms
    }
    
    func getCurrentSpeed() -> WorldSpeed {
        return worldsClock.getCurrentSpeed();
    }
    
    // accauntant
    func calculateIncome(inCome: Int) {
        dayRevenue = inCome
        monthRevenue += inCome
        balance += inCome
        //log
    }
    
    func calculateOutCome() {
        var outCome = 0
        switch service {
        case .normal:
            outCome += normalCost
        case .medium:
            outCome += mediumCost
        case .expensive:
            outCome += expensiveConst
        }
        
        //        outcome += rooms.filter(isRoomOccupiedByCustomer).map { room in
        //            switch room.roomType {
        //            case .president(cost, _): cost
        //            case .econom(cost, _): cost
        //            case .suite(cost, _) : cost
        //            }
        //        }
        let roomsInUse = rooms.filter({ room in
            room.isRoomOccupiedByCustomer()
        }).count
        
        outCome += roomsInUse * 20 + rooms.count - roomsInUse
        balance -= outCome
    }
    
    func payTaxes() {
        let tax = Double(monthRevenue) * 0.1
        balance -= Int(tax)
        monthRevenue = 0
        //log
    }
    
    func cleanRooms() {
        let roomsWaitingClean = rooms.filter { $0.isRoomWaitingCleaing() }
        let capability: Int
        switch service {
        case .normal:
            capability = normalCapability
        case .medium: capability = mediumCapability
        case .expensive: capability = expensiveCapability
        }
        
        let indexes = roomsWaitingClean.prefix(capability).map { room in
            room.room.number
        }
        for i in indexes {
            rooms[i].roomBusiness = EmptyRoom()
        }
        
    }
    
    func checkoutCustomers() {
        var checkedOutCustomers = 0
        rooms.forEach { room in
            if room.isRoomOccupiedByCustomer() {
                if (room.roomBusiness as! RoomInUse).endDate == worldsClock.getDate() {
                    checkedOutCustomers+=1
                    rooms[room.room.number].roomBusiness = DirtyRoom()
                }
            }
        }
        //log count
    }
    
    func checkInPersons(persons: Array<Person>) {
        let roomsByType = Dictionary(grouping: rooms.filter{
            $0.isRoomReadyForCustomer()
        }) { room in
            room.room.roomType
        }.sorted { first, second in
            let types = [RoomType.econom(40, 1500),RoomType.suite(95, 2500), RoomType.president(150, 3500), ]
            let indexFisrst = types.firstIndex(of: first.key)!
            let indexSecond = types.firstIndex(of: second.key)!
            return indexFisrst > indexSecond
        }
        
        
        var income = 0
        roomsByType.forEach { roomByType in
            let personPreferences = persons.filter { person in
                person.roomPreferences.contains(roomByType.key)
            }
            let freeRoomsByType = roomByType.value
            if !freeRoomsByType.isEmpty && !personPreferences.isEmpty {
                let personToCheckin = personPreferences.prefix(freeRoomsByType.count)
                for i in 0...personToCheckin.count {
                    let roomCost = getRoomCost(type: freeRoomsByType[i].room.roomType)
                    
                    income += Int(Double(roomCost)*worldsClock.getMultiply()) * personToCheckin[i].daysAmount
                    var room = rooms[freeRoomsByType[i].room.number]
                    room.roomBusiness = CustomerRoom(startDate: personToCheckin[i].startDate, endDate: personToCheckin[i].endDate)
                    let event = BreakeEvent.generateEvent(person: personToCheckin[i])
                    if event != nil {
                        if event!.completelyBreak {
                            room.lifeTime = 0
                        }else {
                            balance -= event!.costToRepair
                            room.lifeTime -= event!.daysDecrement
                        }
                    }
                }
            }
        }
        calculateIncome(inCome: income)
    }
    
    private func getRoomCost(type: RoomType) -> Int {
        switch type {
        case let .econom(cost, _) : return cost
        case let .suite(cost, _): return cost
        case let .president(cost, _): return cost
        }
    }
    
    func checkDesk() {
        decreaseRoomLifeTime()
        cleanRooms()
        calculateOutCome()
        checkoutCustomers()
        renewRoomsAfterRepairing()
        checkRoomsNeedRepairing()
    }
    
    func checkRoomsNeedRepairing() {
        let needRepair = rooms.filter {
            $0.lifeTime < 1 && !$0.isRoomOccupiedByCustomer()
        }
        needRepair.forEach(repairRoom)
    }
    
    func renewRoomsAfterRepairing() {
        let renewedRooms =
        rooms.filter { $0.isRoomRepairing() && ($0.roomBusiness as! RoomInUse).endDate == worldsClock.getDate() }
        
        renewedRooms.forEach {
            rooms[$0.room.number].roomBusiness = EmptyRoom()
            rooms[$0.room.number].lifeTime = 100
        }
    }
    
    func decreaseRoomLifeTime() {
        rooms.filter {
            $0.isRoomOccupiedByCustomer()
        }.forEach {
            rooms[$0.room.number].lifeTime -= 1
        }
    }
    
    func repairRoom(room: HotelRoom) {
        balance -= getRoomCost(type: room.room.roomType)
        var dayComponent = DateComponents()
        dayComponent.day = 1
        let date = Calendar.current.date(byAdding: dayComponent , to: getToday())!
        rooms[room.room.number].roomBusiness = RepairingRoom(startDate: worldsClock.getDate(), endDate: date)
    }
}
